TECHDEBT.md

```diff
@@ -0,0 +1,78 @@
+# Tech debt
+
+### File Conversion Service
+Observed: January 2021
+Impact: (if this tech debt affected your work somehow, add a +1 here with a
+date and optionally a note)
++1 Christopher 2021 Jan 15 — This is an example of a +1
+##### Problem:
+The current file conversion service for Explorer (converts JSON to TSV/CSV for download)
+is located on the frontend. Ideally, we'd like to place this on the server side, however,
+Graphql does not play nice with non-object data types. The conversion service resolves the requested
+file type to a *string*, which creates some Graphql issues. You might ask, just return the string in a JSON?
+Returning the data is not the *direct* issue. We can resolve to a string and return
+the data from the API back to the user for download no problem. The issue arises because
+we have the Query page in portal that allows users to interact with Graphql (via Guppy), and
+if we were to simply return a JSON, Graphql will populate all of the user's requested query field
+properties as null. This is not ideal. Small example of this requesting with CSV format:
+```gql
+query {
+    subject(format: CSV) {
+        name
+        gender
+    }
+}
+```
+```js
+{
+    "data": {
+        "subject": [
+            {
+                "name": null,
+                "gender": null
+            }
+        ]
+    }
+}
+```
+When really, the conversion service (when moved to the server side) returned:
+```js
+{
+    "data": {
+        "subject": ['name,gender/r/n4xotff,unknown/r/nhzv1z,male']
```

Comment:
Is this the actual thing returned by the conversion service? I think the reason why you got `null` is because the conversion service in here actually returned an array rather than a string, but the fact is that neither field `name` or `gender` exists in the array config index. So there is a type mismatch wrt to apollo's prespective.

If it did returned an array, then probably stripping out the string from array would work. But then we will need to worry about parsing the string etc... 🤪 So I'm fine with keeping this as is. But you can add this info into this tech debt doc if you like

Reply:
No, so the conversion service is _not_ returning null. The issue arises in the Query page (where the user plays around with Guppy) that will eventually resolve to null if we were to leave this in the server. Think of the fields a user submits in a given Query, in this case, `name` and `gender`. These fields would be null in the Query page. It's described in detail in TECHDEBT.MD.

Reply:
yeah I'm not saying the conversion service is returning `null`, I'm saying if it is returning an array rather than a string, then that probably is the cause of this problem

Reply:
If we were to just return a string that would not fix the problem either. You actually need the array in this case based on the Query schema. The issue is because there all no fields matching the user's Query request in the actual response, so those fields, `name` and `gender`, would be shown as null even though there really is data being returned.

Reply:
hmmm yeah I see. Thanks for the explaination